### ISSUE TITLE
Migrate to "Expo modules", from `react-native-unimodules`

### ISSUE THREAD

**Issue by chrisbobbe** [zulip/zulip-mobile#5133]
> The `react-native-unimodules` package is deprecated as of SDK 43, and the module system and autolinking implementation now live in the expo package instead, starting with `expo@43.0.0-beta`.

—https://blog.expo.dev/whats-new-in-expo-modules-infrastructure-7a7cdda81ebc

The migration guide is [here](https://github.com/expo/fyi/blob/main/expo-modules-migration.md).

We'll want to do the "manual configuration" step because we have nontrivial custom code in `android/` and `ios/` and can't just clear those out and run a command to replace them.

**Comment by chrisbobbe** [comment]
OK, I've poked around a bit, trying to understand the changes listed [here](https://github.com/expo/fyi/blob/main/expo-modules-migration.md#manual-configuration) (with instructions to "apply" them to our project, but nothing that helps us understand exactly what'll happen if we do).

My pretty good guess right now is that those diffs are a subset of the changes in a "template app" maintained by Expo: the `templates/expo-template-bare-minimum` directory in https://github.com/expo/expo, between the `sdk-42` branch and the `sdk-43` branch, i.e.,

```
git diff upstream/sdk-42..upstream/sdk-43 -- templates/expo-template-bare-minimum
```

I wanted to know what commits led to those particular changes; here they are (it's not a huge list):

```
$ git lol upstream/sdk-42..upstream/sdk-43 -- templates/expo-template-bare-minimum
6877c1f5c [templates] Bump up expo-updates to 0.10.15 (#15257)
282f5070b [templates] Add missing react test render and upgrade jest (#14993)
2726b3298 [templates] Publish
c75c6c417 [template] Fix hermes command not found on monorepo (#15154)
87a3d02aa [android] Fix building error from outside of project root (#15109)
231fd1415 [templates] Publish
afa47a57b Update template versions
d034a6761 [react-native] Update from 0.64.2 -> 0.64.3 (#15108)
03ad7ee61 Revert "[template] Add back updates manifest embed script"
3ab513775 [template] Add back updates manifest embed script
5b0c17a32 [templates] Update templates
d13d80c8f [template] Bump version for expo run changes
ffdcd2a62 [template] Use expo run commands in bare templates
6e90b5214 [templates] Bump some dependencies
c026bda1d [templates] Publish updates for latest package versions
3de590bb7 [template] Update bare template
8e1d7c05b [template] Set PROJECT_ROOT based on xcode project to fix release builds in monorepo
7946cd428 [template] Update bare template
3bb65c866 [template] Use symbol to access reactNativePath rather than string (which does not work)
f1812e55b [android] Added support for opening https links (#14638)
7ba6eeb78 [templates] Update for latest publish packages
c152aa35c Update project templates to expo@43.0.0-beta.2
d3b29ead0 [templates] Automatically resolve RN path in Xcode build scripts (#14546)
38acbdec4 Update versions in project templates
9886c0aaf [expo-modules][ios] Fix errors from use_frameworks in Podfile (#14523)
9781212eb [templates] Update templates for the next SDK (#14409)
505f2066b [expo] Updates splash, adds dist/ to gitignore (#14439)
eb2c86e38 [ios] Fix versioning script and shell app workflows (#14436)
09e76aae3 [templates] Remove expo forked react-native from managed apps (#14261)
dbd384b22 [template] React Native Image gif and webp support (#14163)
d3f983699 [ios] fix autolink error from pod install (#14034)
507e608fd [ios][autolinking] Move autolinking scripts from the adapter to expo-modules-core (#14046)
896103abd [sdk] upgrade typescript to v4.3.5 (#14015)
f05c8ccb6 [android] Update Gradle to 6.9 (#13929)
05f2ed0c7 [autolinking] Use NodeJS to resolve autolinking script location & popularise it among other imports (#13901)
9e66f3bce [template] Add react-native-safe-area-context to expo-template-bare-minimum (#13840)
3ad2d33e8 Added iOS background color support (#13607)
060fbd4fa Remove xib file (#13590)
b9f2132e7 Remove AppleTVOS12 JavaScriptCore.framework (#13591)
79ca9839c [expo-constants] Modularized  expo-constants without further app setup (#13424)
15aa41994 Update AndroidManifest.xml (#13173)
c44d9daa8 Remove READ_PHONE_STATE permission from default template (#13176)
63b5a92b8 [templates] Remove unnecessary imports
```

I think we can do basically what we've been doing for the last few React Native upgrades: study each of these commits to gather tasks we want to do on our own project (ignoring changes that are badly reasoned, reverted later, or just not relevant to our project in its current state), then do those tasks in our own disciplined way with well-ordered, clear, coherent commits.

I said above that the changes dumped into the migration guide are a subset of the diff I indicated…maybe we'll find out if some changes are intentionally omitted, and if we want to ignore them? The following files are touched in the diff but don't get mentioned in the migration guide:
- templates/expo-template-bare-minimum/android/app/src/main/AndroidManifest.xml
- templates/expo-template-bare-minimum/android/gradle/wrapper/gradle-wrapper.properties
- templates/expo-template-bare-minimum/gitignore
- templates/expo-template-bare-minimum/ios/HelloWorld/Base.lproj/LaunchScreen.xib
- templates/expo-template-bare-minimum/yarn.lock
- templates/expo-template-bare-minimum/ios/Podfile.lock

Our `yarn.lock` and `ios/Podfile.lock` (corresponding to the last two in that list) will be changed automatically when we run `yarn` after changing our `package.json` and/or `ios/Podfile`. We should check those automatic changes against what we see in this diff.

And the following files are mentioned, but in broad strokes that don't capture everything that appears in the diff:
- templates/expo-template-bare-minimum/ios/HelloWorld.xcodeproj/project.pbxproj
- templates/expo-template-bare-minimum/package.json

So we'll want to look into those more closely.

**Comment by chrisbobbe** [comment]
This'll be too long for a commit message, so I'll leave it here. We have five `expo-*` modules, and they vary in whether they cut a new major version for the breaking change from "Unimodules" to "Expo modules". (As one effect of this, I haven't bothered to send one of my easy, catch-all `yarn upgrade` PRs in a long time. To get it mergeable, I'd have had to spend time figuring out exactly which Expo packages were violating semver, and how, and exclude them. That seemed annoying.)

Collapsed below are the details of how I plan to upgrade those five modules, in the same commit that switches us over from `react-native-unimodules` to `expo`.

<details>
<summary>`expo-web-browser`</summary>

expo-web-browser seemed to make the switch at 9.3.0:
- Migrated in expo/expo@314e850c3 and expo/expo@8285c0328.
- 9.3.0 was likely published from expo/expo@fe74edd44. (There's no Git tag, which I think is the convention.)
- The [changelog](https://github.com/expo/expo/blob/master/packages/expo-web-browser/CHANGELOG.md#930--2021-09-09) doesn't mention any breaking changes we'd encounter by upgrading there. So, upgrade to ^9.3.0 with no further changes.
</details>

<details>
<summary>`expo-application`</summary>

expo-application seemed to make the switch at 3.3.0:
- Migrated in expo/expo@8285c0328 and expo/expo@efd75decd.
- 3.3.0 was likely published from expo/expo@fe23a511e.
- The [changelog](https://github.com/expo/expo/blob/master/packages/expo-application/CHANGELOG.md#330--2021-09-08) doesn't mention any breaking changes we'd encounter by upgrading there. So, upgrade to ^3.3.0 with no further changes.
</details>

<details>
<summary>`expo-apple-authentication`</summary>

expo-apple-authentication seemed to make the switch at 4.0.0:
- Migrated in expo/expo@314e850c3 and expo/expo@8285c0328
- 4.0.0 was likely published from expo/expo@1fffde734.
- The [changelog](https://github.com/expo/expo/blob/master/packages/expo-apple-authentication/CHANGELOG.md#400--2021-09-28) only notes one breaking change we'd encounter by upgrading to there: iOS 11 support is dropped. We've already done that.
- The changelog notes a few changes to align the types to the actual implementation, from expo/expo#13567. So, propagate those (really jsdoc plus TODOs for actual meaningful type changes) to our libdef, and upgrade to ^4.0.0.
</details>

<details>
<summary>`expo-sqlite`</summary>

expo-sqlite seemed to make the switch at 10.0.0:
- Migrated in expo/expo@8285c0328 and expo/expo@efd75decd
- 10.0.0 was likely published from expo/expo@1fffde734.
- The [changelog](https://github.com/expo/expo/blob/master/packages/expo-sqlite/CHANGELOG.md#1000--2021-09-28) only notes one breaking change we'd encounter by upgrading to there: iOS 11 support is dropped. We've already done that.
- The changelog notes a small change to the types (expo/expo#13826), which we already took in our 995422ae4. So, upgrade to ^10.0.0 with no further changes.
</details>

<details>
<summary>`expo-screen-orientation`</summary>

expo-screen-orientation seemed to make the switch at 4.0.0:
- Migrated in expo/expo@8285c0328, expo/expo@efd75decd, and expo/expo@6e35719a7.
- 4.0.0 was likely published from expo/expo@1fffde734.
- The [changelog](https://github.com/expo/expo/blob/master/packages/expo-screen-orientation/CHANGELOG.md#400--2021-09-28) only notes one breaking change we'd encounter by upgrading to there: iOS 11 support is dropped. We've already done that.
- The changelog notes a small change to the types (expo/expo#14150): an existing type is exported, and a lot of jsdocs are added. So, follow that.
- We find that there's a breaking change at 4.1.0 (likely published from expo/expo@df68325f3), so we exclude that from the range. In particular, expo/expo@057488d1a started depending on something called `ExpoReactDelegateHandler` offered by expo-modules-core, which wasn't available until expo/expo@0f5f192f4 landed in expo-modules-core@0.6.0 (likely also published from expo/expo@df68325f3).
  - expo-screen-orientation@4.1.0 doesn't seem to enforce expo-modules-core@0.6.0 or later with (peer-)dependency requirements; it probably should. (There don't seem to be (peer-)dependency requirements that *disallow* it, which is good.)
  - I think we're not ready to take expo-modules-core@0.6.0; I suspect (though unverified) that we may need to do a React Native upgrade before that. When I tried to bring that in, I got this failure from `pod install`:
    
    ```
    Installing ExpoModulesCore 0.6.4 (was 0.4.10)
    [!] The following Swift pods cannot yet be integrated as static libraries:
    
    The Swift pod `ExpoModulesCore` depends upon `ReactCommon`, which does not define modules. To opt into those targets generating module maps (which is necessary to import them from Swift when building as static libraries), you may set `use_modular_headers!` globally in your Podfile, or specify `:modular_headers => true` for particular dependencies.
    ```
- So, upgrade to >=4.0.0 <4.1.0.
</details>
